src/dbnode/storage/wide/wide_query_iter_prop_test.go

```diff
@@ -0,0 +1,224 @@
+// +build big
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package wide
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/ts"
+	xtime "github.com/m3db/m3/src/x/time"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/leanovate/gopter"
+	"github.com/leanovate/gopter/gen"
+	"github.com/leanovate/gopter/prop"
+)
+
+type expectedRecord struct {
+	vals         []float64
+	expectedTags []byte
+	expectedID   []byte
+}
+
+type testRecord struct {
+	record   ShardIteratorRecord
+	expected expectedRecord
+}
+
+func genRecord(encoderPool encoding.EncoderPool, start time.Time) gopter.Gen {
+	return gopter.CombineGens(
+		gen.SliceOf(gen.Float64()),
+		gen.AnyString(),
+		gen.AnyString(),
+	).Map(func(val []interface{}) testRecord {
+		var (
+			vals = val[0].([]float64)
+			id   = val[1].(string)
+			tags = val[2].(string)
+		)
+
+		encoder := encoderPool.Get()
+		encoder.Reset(start, len(vals), nil)
+		for i, v := range vals {
+			timestamp := start.Add(time.Duration(i+1) * time.Second)
+			dp := ts.Datapoint{
+				Value:          v,
+				Timestamp:      timestamp,
+				TimestampNanos: xtime.ToUnixNano(timestamp),
+			}
+
+			encoder.Encode(dp, xtime.Nanosecond, nil)
+		}
+
+		seg := encoder.Discard()
+		var encodedData []byte
+		if seg.Head != nil {
+			seg.Head.IncRef()
+			encodedData = append(encodedData, seg.Head.Bytes()...)
+			seg.Head.DecRef()
+		}
+
+		if seg.Tail != nil {
+			seg.Tail.IncRef()
+			encodedData = append(encodedData, seg.Tail.Bytes()...)
+			seg.Tail.DecRef()
+		}
```

Comment:
Nit: might just skip those Inc/DecRef in test code?
Probably needs `seg.Finalize()` if you want to get the full life cycle.